infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -175,231 +186,272 @@ public void handleTailRequest(CorfuPayloadMsg<TailsRequest> msg, ChannelHandlerC
+            responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            r.sendResponse(getResponseMsg(responseHeader, getOverwriteErrorMsg(owe.getOverWriteCause().getId())), ctx);
+        } else if (ex.getCause() instanceof  DataOutrankedException) {
+            responseHeader = getHeaderMsg(req.getHeader(), false, false);
+            r.sendResponse(getResponseMsg(responseHeader, getDataOutrankedErrorMsg()), ctx);
+        } else if (ex.getCause() instanceof  ValueAdoptedException) {
             ValueAdoptedException vae = (ValueAdoptedException) ex.getCause();
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_VALUE_ADOPTED.payloadMsg(vae.getReadResponse()));
+            responseHeader = getHeaderMsg(req.getHeader(), false, false);
+            r.sendResponse(getResponseMsg(responseHeader, getValueAdoptedErrorMsg(vae.getReadResponse())), ctx);
         } else if (ex.getCause() instanceof TrimmedException) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_TRIMMED.msg());
+            responseHeader = getHeaderMsg(req.getHeader(), false, false);
+            r.sendResponse(getResponseMsg(responseHeader, getTrimmedErrorMsg()), ctx);
         } else {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_SERVER_EXCEPTION.payloadMsg(new ExceptionMsg(ex)));
+            responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            r.sendResponse(getResponseMsg(responseHeader, getUnknownErrorMsg(ex)), ctx);
             throw new LogUnitException(ex);
         }
     }
 
     /**
      * Service an incoming write request.
      */
-    @ServerHandler(type = CorfuMsgType.WRITE)
-    public void write(CorfuPayloadMsg<WriteRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
-        LogData logData = (LogData) msg.getPayload().getData();
-        log.debug("log write: type: {}, address: {}, streams: {}", logData.getType(),
-                logData.getToken(), logData.getBackpointerMap());
+    @RequestHandler(type = PayloadCase.WRITE_LOG_REQUEST)
+    public void handleWrite(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        LogData logData = getLogData(req.getPayload().getWriteLogRequest().getLogData());
+        log.debug("handleWrite: type: {}, address: {}, streams: {}",
+                logData.getType(), logData.getToken(), logData.getBackpointerMap());
 
         // Its not clear that making all holes high priority is the right thing to do, but since
         // some reads will block until a hole is filled this is required (i.e. bypass quota checks)
         // because the requirement is to allow reads, but only block writes once the quota is exhausted
         if (logData.isHole()) {
-            msg.setPriorityLevel(PriorityLevel.HIGH);
+            req = getRequestMsg(getHighPriorityHeaderMsg(req.getHeader()), req.getPayload());
         }
 
-        batchWriter
-                .addTask(WRITE, msg)
+        final RequestMsg fReq = req;
```

Comment:
This is strange, why is it needed?

Reply:
I thought it's the effectively final constraint enforced in the lambda expressions.